File: 02 - Lemonade Change/p1.cpp
#include<bits/stdc++.h>
using namespace std;

bool lemonadeChange(vector<int>& bills) {

    int rs5 = 0;
    int rs10 = 0;
    int n = bills.size();

    for( int i = 0; i < n; i++){
        if( bills[i] == 5){
            rs5++;
        }
        else if( bills[i] == 10){
            if( rs5  > 0){
                rs5--;
            }else{
                return false;
            }
            rs10++;
        }
        else{
            if( (rs5  > 0) &&( rs10 > 0)){
                rs5--;
                rs10--;
            }else if( rs5 >= 3){
                rs5 -= 3;
            }
            else{
                return false;
            }
        }
    }

    return true;
        
}

int main(){

    vector<int> bills {5,5,20,5,5,10,5,10,5,20};
    int ans = lemonadeChange(bills);

    cout<<"lemonadeChange : "<<ans<<endl;

    return 0;

}

File: 02 - Lemonade Change/p1.py
def lemonadeChange( bills):
        """
        :type bills: List[int]
        :rtype: bool
        """
        rs5 = 0
        rs10 = 0

        for cost in bills:
            if cost == 5:
                rs5 += 1
            elif cost == 10:
                if rs5 >= 1 :
                    rs10 += 1
                    rs5 -= 1
                else :
                    return False
            else:
                if rs5 >= 1 and rs10 >= 1:
                    rs5 -= 1
                    rs10 -= 1
                elif rs5 >= 3:
                    rs5 -= 3
                else:
                    return False
        
        return True
    
if __name__=='__main__':
    bills = [5,5,20,5,5,10,5,10,5,20]
    print(" bill : ", lemonadeChange(bills))